nomos-core/chain-defs/src/mantle/ledger.rs

```diff
@@ -0,0 +1,159 @@
+use bytes::{Bytes, BytesMut};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    crypto::Digest as _,
+    mantle::{keys::PublicKey, tx::TxHash, Transaction, TransactionHasher},
+    utils::serde_bytes_newtype,
+};
+
+pub type Value = u64;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct NoteId(pub [u8; 32]);
+
+serde_bytes_newtype!(NoteId, 32);
+
+impl NoteId {
+    #[must_use]
+    pub const fn as_bytes(&self) -> &[u8; 32] {
+        &self.0
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
+pub struct Note {
+    pub value: Value,
+    pub pk: PublicKey,
+}
+
+impl Note {
+    #[must_use]
+    pub const fn new(value: Value, pk: PublicKey) -> Self {
+        Self { value, pk }
+    }
+
+    #[must_use]
+    pub fn to_bytes(&self) -> [u8; 40] {
+        let mut bytes = [0u8; 40];
+        bytes[..8].copy_from_slice(&self.value.to_le_bytes());
+        bytes[8..].copy_from_slice(&self.pk.as_bytes()[..]);
+        bytes
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Tx {
+    pub inputs: Vec<NoteId>,
+    pub outputs: Vec<Note>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Utxo {
+    pub tx_hash: TxHash,
+    pub output_index: usize,
+    pub note: Note,
+}
+
+impl Utxo {
+    #[must_use]
+    pub fn id(&self) -> NoteId {
+        let mut hasher = crate::crypto::Hasher::new();
+
+        hasher.update(b"NOMOS_UTXO_ID_V1");
```

Comment:
Nipick: Maybe an outside const def? It can be in Utxo level:

```suggestion
impl Utxo {
    const UTXO_ID: &[u8] = &b"NOMOS_UTXO_ID_V1";
    #[must_use]
    pub fn id(&self) -> NoteId {
        let mut hasher = crate::crypto::Hasher::new();
        hasher.update(Self::UTXO_ID);
        ...
```

Use better naming, utxo_id is not really a good suggestion maybe.